#962 Java 8 Stream conversions to Collections

filiphr opened the pull request:
This PR covers the easier part of #962 (converting a `Stream` to some collection type).  The implementation type are same as for the [collection mappings](http://mapstruct.org/documentation/1.1/reference/html/index.html#implementation-types-for-collection-mappings) (without the `Map` implementations of course 😄)

Documentation is not updated, once this is resolved I will update the documentation.

Also I am not sure why, but in the generated mappers there are no null checks for the Stream conversions. What am I doing wrong?

sjaakd commented:
> Also I am not sure why, but in the generated mappers there are no null checks for the Stream conversions. What am I doing wrong?

Hmm. That should be included automatically. You might want to single-step through `assignToPlain`.  Although you might argue that `Optional.ofNillable` might be more in line with Java8. :smile:

filiphr commented:
@sjaakd Thanks for the pointer. However, the problem is not there it is in `assignToCollection`. As the `targetType` is `Collection` the entire conversion is `null` checked but not the `<SOURCE>` in the conversion.

Example:
```java
List<String> list = source.getStringStream().collect( Collectors.toList() );
if ( list != null ) {
    target.setStringStream( list );
}
```

The `<SOURCE>` in the previous snippet is `source.getStringStream()`.

I also tried with `NullValueCheckStrategy.ALWAYS` and same result. Is this expected or is it a bug?

> Although you might argue that Optional.ofNillable might be more in line with Java8. 😄

I agree, but I think that I shouldn't do a `null` check in the conversion itself?

sjaakd commented:
> As the targetType is Collection the entire conversion is null checked but not the <SOURCE> in the conversion.

ahh.. now I get it.. This (PR) is the first time we have a type conversion from collection to collection type. I believe it is (or at least was) even in the comments somewhere. I refactored these particular parts some commits ago.

For mapping methods, the responsibility for null checking is in the (forged) mapping method. 

So perhaps its not such a bad idea to do it in the conversion in Java8 style. Form experience: the placing of the null check is the result of a long discussion (see #956 for instance) and a bit of a can of worms. 

BTW: I'm still busy with refactoring this whole property mapping (step-by-step). I think it has become far to complex / nontransparent.

filiphr commented:
Done. Using Optional it looks much better now, thanks :smile:. The only thing is that, it is in one line and it is a bit long, I am not sure if it is possible to do it multi-line and to look nice.

sjaakd commented on processor/src/main/java/org/mapstruct/ap/internal/conversion/JavaStreamToCollectionConversion.java:
```diff
@@ -41,7 +41,9 @@
     @Override
     protected String getToExpression(ConversionContext conversionContext) {
-        return "<SOURCE>.collect( Collectors." + collector( conversionContext ) + " )";
+        return "Optional.ofNullable( <SOURCE> )" +
```
could simply adding enters not be an option

filiphr replied:
It doesn't help. The `TypeConversion.ftl` has `compress single_line = true` which leads to making the `\n` into spaces (At least I think it is because of that)

sjaakd replied:
ah.. we should fix that someday. I'll try to take that on when I do the refactoring on the templates for the property mapping. For now.. leave it.

agudian replied:
See #877 🙄 😄

sjaakd commented:
> Using Optional it looks much better now, thanks 😄

No problem.. I'll have a thorough look tomorrow. Thanks.

gunnarmorling commented on processor/src/main/java/org/mapstruct/ap/internal/conversion/JavaStreamToCollectionConversion.java:
```diff
@@ -0,0 +1,95 @@
+            collector = "toCollection( TreeSet::new )";
+        }
+
+        //TODO what if we register wrong. In the end it won't compile, but should we throw an exception?
```
Yes, we should never generate code that doesn't compile. Instead we should raise a descriptive error about the issue before (missing conversion etc.).

filiphr replied:
We will already raise such error, as no conversion could be found before. The only reason for something like this to happen is if we register a new type and forget to update the `JavaStreamToCollectionConversion`. In order to avoid such things, we could maybe do something in the register methods and make sure that the providers supports the conversion that we are trying to register.

gunnarmorling commented on processor/src/test/java/org/mapstruct/ap/test/conversion/java8stream/Source.java:
```diff
@@ -0,0 +1,109 @@
+
+    private List<Integer> ints;
+
+    private Stream<Integer> stream;
```
Do we have a test that invokes a method for converting the element type?

sjaakd replied:
Indeed. I'm a bit worried on the erasure. Possible problems would show.

filiphr replied:
The erasure is a problem as it will allow a `List<String>` to be converted from/to `Stream<Integer>`. I think that we need to do something like in the `BuiltInMethod`(s) there the generics are handled nice.  Maybe even use them instead

filiphr replied:
With my latest commits the approach is changed and we have a test that maps `Stream<String>` to `List<Integer>`.

gunnarmorling replied:
Does it also work by invoking another mapping method for the element type now? Or is that left for a follow-up PR?

filiphr replied:
Yes, with the new approach it takes the same approach ad the `IterableMappingMethod` and checks for a mapping method for the element type and uses it. I will add a test for that in any case.

filiphr replied:
@gunnarmorling I forgot that I already added that. Have a look at the `TargetElement` and `SourceElement`, they are mapped via another mapping method.

gunnarmorling commented on processor/src/test/java/org/mapstruct/ap/test/conversion/java8stream/StreamsTest.java:
```diff
@@ -0,0 +1,100 @@
+        target.setIntegerIterable( Arrays.asList( 10, 11, 12 ) );
+        target.setSortedSet( new TreeSet<Integer>( Arrays.asList( 12, 11, 10 ) ) );
+        target.setNavigableSet( new TreeSet<Integer>( Arrays.asList( 12, 11, 10 ) ) );
+
```
The double empty lines look odd (also below).

sjaakd commented on processor/src/main/java/org/mapstruct/ap/internal/conversion/JavaStreamToCollectionConversion.java:
```diff
@@ -0,0 +1,95 @@
+        Type targetType = conversionContext.getTargetType();
+        TypeFactory typeFactory = conversionContext.getTypeFactory();
+        String collector = "";
+        // we must do erasure as JDK compiler does not work without it
```
Hmm. I'm not sure whether this is correct. `Iterable<String>` is not assignable to `Iterable<Object>`.. Is it the JDK compilation cycle that fails or the Eclipse? (Note both are executed)

filiphr replied:
This conversion is only for the same type currently. Without the `erasure()` the JDK compilation fails, and the Eclipse is OK. I think that the problem is that when we do `typeFactory.getType( Iterable.class )` we get a type like `Iterable<T>`. I am not sure what exactly is correct and how to check if the `targetType` is `Iterable`, or `Set`, or something else.

sjaakd replied:
@agudian .. could it  be that we need the `declared type` iso`type` comparison.?

sjaakd commented on processor/src/main/java/org/mapstruct/ap/internal/conversion/Conversions.java:
```diff
@@ -352,15 +366,15 @@ public boolean equals(Object obj) {
                 }
             }
-            else if ( !sourceType.equals( other.sourceType ) ) {
+            else if ( !sourceType.erasure().equals( other.sourceType.erasure() ) ) {
```
why is this needed?

filiphr replied:
This is another point where I was not sure. If they are removed then the conversions are not found.
It seems that the type `Iterable<T>` is not the same to `Iterable<T>` (I don't know why)

sjaakd commented on processor/src/main/java/org/mapstruct/ap/internal/conversion/JavaStreamToCollectionConversion.java:
```diff
@@ -0,0 +1,95 @@
+ *
+ * @author Filip Hrisafov
+ */
+public class JavaStreamToCollectionConversion extends SimpleConversion {
```
perhaps we should work with an abstract `JavaStreamToCollectionConversionBase` which has specific implementations for `Iterable, `Collection`, `Set`.. etc.  Then you could make the method `collector` abstract and removing the need for the current collector. This is done for other conversions as well (harmonize) and moves this if-elseif-elseif to the `Conversions`.

filiphr commented:
@gunnarmorling, @agudian, @sjaakd I rethought the approach about the implementation of this feature and I went in the same direction as the `IterableMappingMethod`. Should I just update this PR or you would prefer a different PR with the new approach?

The new approach works with mapping the element type considerts the different mappings:
* `Iterable` or `Array` type -> Stream type
* `Stream<TypeA>` to `Stream<TypeB>`
* `Stream` -> `Iterable` or `Array` type

filiphr commented:
I referenced a wrong issue with my last commit. 

I changed to approach for the stream mappings to be inline with the `IterableMappingMethod`. Mapping between different types work by using the stream `map` method.

I think that there are some formatting issues in the generated mappers that need to be resolved.

gunnarmorling commented on processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java:
```diff
@@ -836,4 +848,35 @@ public boolean hasEmptyAccessibleContructor() {
+    public Collection<Type> determineTypeArguments(String superClassName) {
+        Class<?> superClass = null;
+        try {
+            superClass = Class.forName( superClassName );
```
What classes could we see here? If it's classes from the application under complication, `Class#forName()` should be avoided. `Elements#getTypeElement()` would be an alternative.

filiphr replied:
I needed to get the type element of the `java.util.stream.Stream` being mapped. I can try the `Elements#getTypeElement()` to see if it works like that. I didn't know about that. I can't call our `determineTypeArguments(Class<?> superClass)` as the `Stream` does not compile with Java 6 and 7

filiphr replied:
@gunnarmorling I had a look at the method that we were using before, and the `Class` was only needed to get it's name. Therefore, I did some changes in my method and moved checks there.

gunnarmorling commented:
@filiphr Are you going to squash your commits into one (or a few ones)? E.g. the "removing imports" one seems not worth a separate entry in the commit history. Also can you prefix all commits that are going to remain with the issue number?

filiphr commented:
@gunnarmorling yes of course. I have to rebase everything as there is some conflict with the `TypeFactory`. We can also do squash and merge directly from the PR. I thought that it is easier for reviewing if I don't squash all the time.

filiphr commented:
@gunnarmorling I was thinking to have one commit. The commit with the issue 971 is with a wrong issue referenced.

sjaakd commented:
@filiphr What's the status on this one?

filiphr commented:
@sjaakd I was waiting for the new PRs to be merged. I will rebase this on top of the latest master. @gunnarmorling said that will have a look at it as well.

filiphr commented:
I squashed and rebased on the latest master branch. @gunnarmorling can you have a look at it?

gunnarmorling commented:
Thanks! Yes, I'll take another look tomorrow.

gunnarmorling commented on processor/src/test/java/org/mapstruct/ap/test/java8stream/StreamMappingTest.java:
```diff
@@ -0,0 +1,244 @@
+    }
+
+    @Test
+    public void shouldMapList() {
```
Hi @filiphr, I took a look at the generated code. That's what I see:

    protected List<String> stringStreamToStringList(Stream<String> stream) {
        if ( stream == null ) {
            return null;
        }

        List<String> list = new ArrayList<String>();

        list.addAll( stream.map( Function.identity() )
            .collect( Collectors.toCollection( ArrayList<String>::new ) )
        );

        return list;
    }

Why is it using the identity function here? And why is it assigning the result to a new list? I think the following would be better to generate:

    protected List<String> stringStreamToStringList(Stream<String> stream) {
        if ( stream == null ) {
            return null;
        }

        return stream.collect( Collectors.toCollection( ArrayList<String>::new ) );
    }

gunnarmorling replied:
Another one I saw is this:

    protected Stream<String> stringArrayListToStringStream(ArrayList<String> arrayList) {
        if ( arrayList == null ) {
            return null;
        }

        Stream<String> stream = Stream.empty();

        stream = arrayList.stream()
        .map( Function.identity() );

        return stream;
    }

I think we could avoid the assignment of `Stream.empty()` which is overridden right in the next line? And the call to `map()` looks not properly indented (I know, it's nit-picking, sorry ;).

filiphr replied:
>Why is it using the identity function here?

Identity is used so I can have a general approach. Will fix this. 

>And why is it assigning the result to a new list?

The assigning to a new list is due to the before/after mapping possible calls. I can try and fix it by adding some checks if there are any before/after calls. Maybe add it as a TODO in #1012?

>I think we could avoid the assignment of Stream.empty() which is overridden right in the next line? 

Same for the `Stream.empty()`.  There could be calls to before/after mapping methods (Although it makes no sense to me to have a before mapping for `Stream`). I can do something similar as for the list.

>And the call to map() looks not properly indented (I know, it's nit-picking, sorry ;).

I am having problems to indent this correctly :(. I am not sure how I can do it correctly, @sjaakd maybe you can help out for this one. It is in line [114](https://github.com/filiphr/mapstruct/blob/java8-stream-conversions/processor/src/main/resources/org/mapstruct/ap/internal/model/StreamMappingMethod.ftl#L114), [117](https://github.com/filiphr/mapstruct/blob/java8-stream-conversions/processor/src/main/resources/org/mapstruct/ap/internal/model/StreamMappingMethod.ftl#L117) and [120](https://github.com/filiphr/mapstruct/blob/java8-stream-conversions/processor/src/main/resources/org/mapstruct/ap/internal/model/StreamMappingMethod.ftl#L120)

gunnarmorling commented on processor/src/test/java/org/mapstruct/ap/test/java8stream/StreamMappingTest.java:
```diff
@@ -0,0 +1,244 @@
+    @Test
+    public void shouldMapListToCollection() {
+        Source source = new Source();
+        source.setIntegerList( Arrays.asList( 1, 2 ).stream() );
```
setIntegerList=setIntegerStream?

filiphr replied:
Will fix this issues in the entire test

gunnarmorling commented on processor/src/test/java/org/mapstruct/ap/test/java8stream/StreamMappingTest.java:
```diff
@@ -0,0 +1,244 @@
+    @Test
+    public void shouldMapIntegerSetToStringSet() {
+        Source source = new Source();
+        source.setAnotherIntegerSet( new HashSet<Integer>( Arrays.asList( 1, 2 ) ).stream() );
```
setAnotherIntegerSet=setAnotherIntegerStream (it seems to be a common issue in this test)

gunnarmorling replied:
Instead of generating

    protected Set<String> integerStreamToStringSet(Stream<Integer> stream) {
        if ( stream == null ) {
            return null;
        }

        Set<String> set = new HashSet<String>();

        set.addAll( stream.map( integer -> String.valueOf( integer ) )
            .collect( Collectors.toCollection( HashSet<String>::new ) )
        );

        return set;
    }

How about this:

    protected Set<String> integerStreamToStringSet(Stream<Integer> stream) {
        if ( stream == null ) {
            return null;
        }

        return stream.map( integer -> String.valueOf( integer ) )
            .collect( Collectors.toCollection( HashSet<String>::new ) )
        );
    }

I.e. avoid the adding to another collection.

filiphr replied:
See my other comments about the same issue

gunnarmorling commented on processor/src/test/java/org/mapstruct/ap/test/java8stream/forged/StreamMappingTest.java:
```diff
@@ -0,0 +1,120 @@
+ */
+@IssueKey("962")
+@RunWith(AnnotationProcessorTestRunner.class)
+public class StreamMappingTest {
```
Could you use a distinct name for this test? There's another one of the same name.

filiphr replied:
Of course. We have the same situation with the `CollectionsMappingTest`. I reused most of the tests in `org.mapstruct.ap.test.collection` 😄

agudian commented:
I usually use `Collectors.toList()` / `Collectors.toSet()` instead of `toCollection( .. )`.

I guess you're that specific here to reuse the same template for the general scenario, right? Including the `map( identity() )` part?

gunnarmorling commented on processor/src/test/java/org/mapstruct/ap/test/java8stream/wildcard/WildCardTest.java:
```diff
@@ -0,0 +1,134 @@
+        Plan.class,
+        Idea.class
+    })
+    public void shouldGenerateExtendsBoundSourceForgedStreamMethod() {
```
Minor thing, I'm seeing a FQN in the generated code here:

    protected List<Plan> ideaStreamToPlanList(Stream<? extends org.mapstruct.ap.test.java8stream.wildcard.Idea> {
        ...
    }

gunnarmorling replied:
(Same for the super bounds test)

filiphr replied:
Will look into it.

filiphr replied:
This is a different issue see #1014

gunnarmorling commented on processor/src/main/java/org/mapstruct/ap/internal/model/StreamMappingMethod.java:
```diff
@@ -0,0 +1,404 @@
+        }
+
+        public StreamMappingMethod build() {
+            //TODO the building of the methods is the same as the IterableMappingMethod, the only difference being
```
Should we track it in a separate issue?

filiphr replied:
Already created one see #1012

gunnarmorling commented on processor/src/main/java/org/mapstruct/ap/internal/model/StreamMappingMethod.java:
```diff
@@ -0,0 +1,404 @@
+        }
+
+        private Assignment forgeMapping(SourceRHS sourceRHS, Type sourceType, Type targetType) {
+
```
Just a small thing: there are some superfluous empty lines like this one here and there.

filiphr replied:
I'll fix those thanks.

gunnarmorling commented on processor/src/main/java/org/mapstruct/ap/internal/model/StreamMappingMethod.java:
```diff
@@ -0,0 +1,404 @@
+            }
+        }
+
+        throw new IllegalStateException( "Method " + this + " has no source parameter." );
```
Is this a "regular" error situation? If so, we should raise a proper compiler error instead.

filiphr replied:
This should never happen. We do the exact same thing in the `IterableMappingMethod`, `MapMappingMethod` and `NestedPropertyMappingMethod`.

gunnarmorling commented on processor/src/main/java/org/mapstruct/ap/internal/model/StreamMappingMethod.java:
```diff
@@ -0,0 +1,404 @@
+        return loopVariableName;
+    }
+
+    public String getDefaultValue() {
```
When is this one needed?

filiphr replied:
In the `StreamMappingMethod.ftl` when we need to reset an array.

gunnarmorling commented on processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java:
```diff
@@ -840,14 +862,27 @@ public boolean hasEmptyAccessibleContructor() {
+
+    /**
+     * Searches for the given superclass and collects all type arguments for the given className.
+     * This is used for support with Java 8 streams, as must use the FQN in order to be able to compile
```
I'm not following why it's using the FQN now. When do we compile with Java 6?

filiphr replied:
This actually works. I only need to use FQN in the `TypeFactory`. I thought that it will fail as we use the FQN in the `Conversions`. However, I ran the integration tests and it seems OK. I will revert the changes to this.

gunnarmorling commented on processor/src/main/java/org/mapstruct/ap/internal/processor/MapperCreationProcessor.java:
```diff
@@ -447,6 +456,30 @@ else if ( method.isEnumMapping() ) {
     }
 
+    private StreamMappingMethod createStreamMappingMethod(SourceMethod method, MappingOptions mappingOptions) {
+        // TODO this is the same as the if with the isIterableMapping, I think that we need
```
Let's create an issue so we don't forget about it.

filiphr replied:
Done see #1012

gunnarmorling commented:
Hey @filiphr, great work! Awesome to have the stream support :) I've added some comments inline, mostly details. Kudos for the super-thorough testing, too!

gunnarmorling commented:
> I usually use Collectors.toList() / Collectors.toSet() instead of toCollection( .. ).
>
> I guess you're that specific here to reuse the same template for the general scenario, right? Including the map( identity() ) part?

I also was about to suggest to use `toList()` etc., but their comments state that there are no guarantees regarding mutability etc. So I think we are better off using the explicit constructor references of specific types.

My guess on the identity function would be the same, would still be nice to avoid if reasonably doable.

filiphr commented:
Thanks for the through review guys.

>I usually use Collectors.toList() / Collectors.toSet() instead of toCollection( .. ).
>
>I guess you're that specific here to reuse the same template for the general scenario, right? Including the map( identity() ) part?

Yes to the both. I wanted to not have too many checks in the model. The `Function.identity()` part can be easily changed. I an do a check on the assignment and if it is a direct assignment do not add the `map`. Regarding the `Collectors.toList()` / `Collectors.toSet() `, I think it is much more complicated to make it like that. If you want we can create a new issue for that and tackle it separately. I think having `Collectors.toCollection( .. )` provides us more flexibility if we want to allow the users to provide custom default implementations.

filiphr commented:
The `map` method is now in the same line as `stream`, `Function.identity()` is not used anymore.

The only thing that is left is:
- [ ] Do we agree to keep using `Collectors.toCollection( .. )`?
- [ ] What about the before/after possible calls?
- [ ] The FQN `org.mapstruct.ap.test.java8stream.wildcard.Idea`

gunnarmorling commented:
> Do we agree to keep using Collectors.toCollection( .. )

Yes, it makes sense to me.

> What about the before/after possible calls?

If possible I'd prefer to avoid the superfluous instantiations in the regular case (no before/after). But it coould be done as a follow-up task if you prefer.

filiphr commented:
>If possible I'd prefer to avoid the superfluous instantiations in the regular case (no before/after). But it coould be done as a follow-up task if you prefer.

I did it as part of this PR. Have a look at my last commit.

gunnarmorling commented:
@filiphr The generated source code looks functionally perfect now. Thanks a lot! The only remaining issue I spotted is a minor formatting glitch:

    return stream.map( string -> Long.parseLong( string ) )
    .collect( Collectors.toCollection( HashSet<Long>::new ) );

Could you indent the second line one level? Other than that it's great, it only needs a rebase (there is one conflict with current master). And do we have the doc update already?

filiphr commented:
Thanks for the review @gunnarmorling.

>Could you indent the second line one level?

For some reason I can't. I would appreciate some help here. You can see the template lines [here](https://github.com/filiphr/mapstruct/blob/java8-stream-conversions/processor/src/main/resources/org/mapstruct/ap/internal/model/StreamMappingMethod.ftl#L120-L121). There is a tab, in Line 121, but once it is included in the nested it does not add it to the template. I can write it in one line and it will look like this:

```java
return stream.map( sourceElement -> mapElement( sourceElement ) ).collect( Collectors.toCollection( ArrayList<TargetElement>::new ) );
```
WDYT?

>And do we have the doc update already?

No, but I will add a commit for it to this PR

gunnarmorling commented:
> For some reason I can't. I would appreciate some help here. 

Ah, I reckon it's due to [IndentationCorrectingWriter](https://github.com/mapstruct/mapstruct/blob/master/processor/src/main/java/org/mapstruct/ap/internal/writer/IndentationCorrectingWriter.java) which re-indents the generated source code at the end solely based on the nesting level as indicated by "{" and "}". It doesn't deal with multi-line chained invocations as we have them here.

If you feel adventureous, you may give it a try to fix it in the writer, but if not let's merge the PR as is and handle it separately. Just let me know what you prefer.

filiphr commented:
>If you feel adventureous, you may give it a try to fix it in the writer, but if not let's merge the PR as is and handle it separately. Just let me know what you prefer.

I had a look at it, I think it deserves as new PR 😄 and I created #1022 for it.

I rebased the PR and added 3 new commits, one for fixing a typo, another one for the documentation and the last one for support of the new `@Context`

agudian commented:
> If you feel adventureous, you may give it a try to fix it in the writer, but if not let's merge the PR as is and handle it separately. Just let me know what you prefer.

Or, we could revisit the idea to simply let a code formatter do the formatting for us, as proposed in #877. A first search didn't yield that many open source libs that seem usable for us there, with the exception of the Eclipse formatter... Not sure on the licensing though, if we start shading it into our processor.jar. But I could try integrating it to see how it behaves...

gunnarmorling commented:
@filiphr This one is good to go, right? Can you rebase, squash as you like and merge it?

filiphr commented:
@gunnarmorling If you are OK with the documentation then it is good to go. I will squash and merge it from github as there are no conflicts and I think that everything belongs into one commit

agudian commented:
better rebase locally, as you'll also have to update the legal headers for the new files... :)

gunnarmorling commented on documentation/src/main/asciidoc/mapping-streams.asciidoc:
```diff
@@ -0,0 +1,67 @@
+The mapping of `java.util.Stream` is done in a similar way as the mapping of collection types, i.e. by defining mapping
+methods with the required source and target types in a mapper interface.
+
+The generated code will contain the creation of a `Stream` from the provided `Iterable`/`Array` or will collect the
```
`Array` should not be marked as literal. It's just an "array".

gunnarmorling commented on documentation/src/main/asciidoc/mapping-streams.asciidoc:
```diff
@@ -0,0 +1,67 @@
+
+The generated code will contain the creation of a `Stream` from the provided `Iterable`/`Array` or will collect the
+provided `Stream` into an `Iterable`/`Array`. If a mapping method or an implicit conversion for the source and target
+element types exists, then this conversion will be done in `Stream.map`. The following shows an example:
```
done via `Stream#map()` (let's use the hash pound when referencing to a method).

gunnarmorling commented on documentation/src/main/asciidoc/mapping-streams.asciidoc:
```diff
@@ -0,0 +1,67 @@
+----
+====
+
+The generated implementation of the `integerStreamToStringSet` performs the conversion from `Integer` to `String` for
```
the `integerStreamToStringSet()` method.

gunnarmorling commented on documentation/src/main/asciidoc/mapping-streams.asciidoc:
```diff
@@ -0,0 +1,67 @@
+
+[WARNING]
+====
+If a mapping from a `Stream` to an `Iterable` or an `Array` is performed, then the passed `Stream` will be consumed
```
`Array`=array

gunnarmorling commented on documentation/src/main/asciidoc/mapping-streams.asciidoc:
```diff
@@ -0,0 +1,67 @@
+[WARNING]
+====
+If a mapping from a `Stream` to an `Iterable` or an `Array` is performed, then the passed `Stream` will be consumed
+and will no longer be possible to consume it.
```
and it

gunnarmorling commented on documentation/src/main/asciidoc/mapping-streams.asciidoc:
```diff
@@ -0,0 +1,67 @@
+and will no longer be possible to consume it.
+====
+
+The same implementation types as in <<implementation-types-for-collection-mappings>>, are used for the creation of the
```
The "," is too much.

filiphr commented:
Thanks for the review guys. I have update the last documentation comments and squashed and merged
